File: backend/src/main/java/com/postwerk/service/AutomationService.java
package com.postwerk.service;

import com.postwerk.dto.*;
import com.postwerk.dto.automation.AutomationValidationResult;
import com.postwerk.model.Automation;
import com.postwerk.model.enums.AutomationKind;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.UUID;
import java.util.function.UnaryOperator;

/**
 * Service interface for managing email automation workflows, including CRUD operations,
 * flow graph updates (nodes and edges), execution history, manual triggering, and bulk import/export.
 *
 * @since 1.0
 */
public interface AutomationService {

    AutomationResponse create(UUID organizationId, UUID actingUserId, AutomationRequest request, String ipAddress);

    List<AutomationResponse> listByOrg(UUID organizationId);

    /** Lists the organization's reusable, trigger-less integrations ({@code kind == INTEGRATION}). */
    List<AutomationResponse> listIntegrations(UUID organizationId);

    AutomationDetailResponse getById(UUID organizationId, UUID automationId);

    AutomationResponse update(UUID organizationId, UUID actingUserId, UUID automationId, AutomationRequest request, String ipAddress);

    void delete(UUID organizationId, UUID actingUserId, UUID automationId, String ipAddress);

    AutomationResponse updateStatus(UUID organizationId, UUID actingUserId, UUID automationId, AutomationStatusRequest request, String ipAddress);

    AutomationDetailResponse updateFlow(UUID organizationId, UUID actingUserId, UUID automationId, FlowUpdateRequest request, String ipAddress);

    AutomationDetailResponse updateConstants(UUID organizationId, UUID actingUserId, UUID automationId, ConstantsUpdateRequest request, String ipAddress);

    Page<AutomationExecutionResponse> getExecutions(UUID organizationId, UUID automationId, Pageable pageable);

    /**
     * Fires an automation's MANUAL trigger on demand. The user-entered {@code parameters} (keyed by the
     * trigger's parameter-set field names) are seeded as {@code trigger.*} execution variables, then the
     * flow runs live via the same off-request path as inbound webhooks. Requires the automation to be
     * ACTIVE and to contain a TRIGGER node in {@code MANUAL} mode.
     */
    void runManually(UUID organizationId, UUID actingUserId, UUID automationId,
                     java.util.Map<String, Object> parameters, String ipAddress);

    List<AutomationExportDto> exportAll(UUID organizationId);

    ImportResultDto importAll(UUID organizationId, UUID actingUserId, List<AutomationExportDto> items, String ipAddress);

    AutomationResponse toggleLock(UUID organizationId, UUID automationId);

    boolean isLocked(UUID organizationId, UUID automationId);

    /**
     * Runs the semantic {@link AutomationValidator} against the automation's saved flow.
     * Used by the on-demand validate endpoint, the AI self-correction loop, and the activate gate.
     */
    AutomationValidationResult validate(UUID organizationId, UUID automationId);

    /**
     * Deep-copies a source automation's flow into a new buyer-owned automation for the marketplace
     * install path. Nodes/edges are recreated; each node's config JSON is transformed by
     * {@code configRewriter} (resource id remapping + account clearing); the raw constants JSONB is
     * copied verbatim so encrypted secrets are preserved. The copy is created PAUSED with no bound
     * accounts; buyer-owned inbound webhook endpoints are reconciled for any WEBHOOK trigger.
     *
     * @param buyerOrgId     the buyer's active organization that will own the copy (#4) — the org the
     *                       install was made from, so its plan/quota governs the copy when it runs
     * @param buyerId        the user that will own the copy (audit/attribution)
     * @param source         the author's source automation entity
     * @param configRewriter transform applied to each node's config JSON
     * @param hidden         whether the copy is content-hidden (PRIVATE listing)
     * @param locked         whether the copy is locked from editing
     * @return the new automation id
     */
    UUID installCopy(UUID buyerOrgId, UUID buyerId, Automation source, UnaryOperator<String> configRewriter,
                     boolean hidden, boolean locked);

    /**
     * Creates a new automation + flow from explicit specs (not from a live source entity). Used to
     * materialize a buyer-owned copy from a marketplace snapshot manifest. Mirrors {@link #installCopy}
     * (PAUSED, no bound accounts, webhook endpoints reconciled) but reads nodes/edges from
     * {@link NodeSpec}/{@link EdgeSpec} lists, so it does not depend on the author's live automation.
     *
     * @return the new automation id
     */
    UUID createSnapshotCopy(UUID organizationId, UUID userId, String name, String description,
                            AutomationKind kind, String color, String constants, String flowData,
                            java.util.List<NodeSpec> nodes, java.util.List<EdgeSpec> edges,
                            boolean hidden, boolean locked);
}

File: frontend/src/app/core/services/automation.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpParams } from '@angular/common/http';
import { ApiService } from './api.service';
import {
  Automation,
  AutomationConstantInput,
  AutomationDetail,
  AutomationExecution,
  AutomationExport,
  AutomationRequest,
  AutomationStatus,
  FlowUpdateRequest,
  TestModeResult,
  TestModeStats,
} from '../../models/automation.model';
import { ImportResult } from '../../models/import-result.model';

/**
 * Provides CRUD operations for email automations, including flow updates,
 * execution history retrieval, manual triggering, and import/export capabilities.
 */
@Injectable({ providedIn: 'root' })
export class AutomationService {
  private readonly api = inject(ApiService);
  private readonly basePath = '/automations';

  list() {
    return this.api.get<Automation[]>(this.basePath);
  }

  /** Lists the user's integrations (trigger-less reusable flows, kind=INTEGRATION). */
  listIntegrations() {
    return this.api.get<Automation[]>(`${this.basePath}/integrations`);
  }

  get(id: string) {
    return this.api.get<AutomationDetail>(`${this.basePath}/${id}`);
  }

  create(request: AutomationRequest) {
    return this.api.post<Automation>(this.basePath, request);
  }

  update(id: string, request: AutomationRequest) {
    return this.api.put<Automation>(`${this.basePath}/${id}`, request);
  }

  delete(id: string) {
    return this.api.delete<void>(`${this.basePath}/${id}`);
  }

  toggleLock(id: string) {
    return this.api.patch<Automation>(`${this.basePath}/${id}/lock`, {});
  }

  updateStatus(id: string, status: AutomationStatus) {
    return this.api.patch<Automation>(`${this.basePath}/${id}/status`, { status });
  }

  updateFlow(id: string, request: FlowUpdateRequest) {
    return this.api.put<AutomationDetail>(`${this.basePath}/${id}/flow`, request);
  }

  updateConstants(id: string, constants: AutomationConstantInput[]) {
    return this.api.put<AutomationDetail>(`${this.basePath}/${id}/constants`, { constants });
  }

  getExecutions(id: string, page = 0, size = 20) {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.api.get<{ content: AutomationExecution[]; totalElements: number }>(
      `${this.basePath}/${id}/executions`,
      { params }
    );
  }

  /**
   * Fires the automation's MANUAL trigger. {@link parameters} are user-entered values keyed by the
   * trigger's parameter-set field names; they are seeded as {@code trigger.*} variables for the run.
   */
  runManually(id: string, parameters?: Record<string, unknown>) {
    return this.api.post<void>(`${this.basePath}/${id}/run`, { parameters: parameters ?? null });
  }

  export() {
    return this.api.get<AutomationExport[]>(`${this.basePath}/export`);
  }

  import(data: AutomationExport[]) {
    return this.api.post<ImportResult>(`${this.basePath}/import`, data);
  }

  // ─── Test Mode ─────────────────────────────────────────

  getTestModeResults(id: string, feedback?: string, page = 0, size = 20) {
    let params = new HttpParams().set('page', page).set('size', size);
    if (feedback) params = params.set('feedback', feedback);
    return this.api.get<{ content: TestModeResult[]; totalElements: number }>(
      `${this.basePath}/${id}/test-mode/results`,
      { params }
    );
  }

  submitTestModeFeedback(id: string, resultId: string, feedback: 'CORRECT' | 'INCORRECT', note?: string) {
    return this.api.patch<TestModeResult>(
      `${this.basePath}/${id}/test-mode/results/${resultId}/feedback`,
      { feedback, note: note || null }
    );
  }

  getTestModeStats(id: string) {
    return this.api.get<TestModeStats>(`${this.basePath}/${id}/test-mode/stats`);
  }

  clearTestModeResults(id: string) {
    return this.api.delete<void>(`${this.basePath}/${id}/test-mode/results`);
  }

  /** Runs one already-synced email through an automation in dry-run and returns the simulated result. */
  simulateEmail(automationId: string, emailId: string) {
    return this.api.post<TestModeResult>(
      `${this.basePath}/${automationId}/test-mode/simulate/${emailId}`,
      {}
    );
  }

  /** Deletes a single test-mode result so it no longer counts toward the accuracy statistics. */
  deleteTestModeResult(automationId: string, resultId: string) {
    return this.api.delete<void>(`${this.basePath}/${automationId}/test-mode/results/${resultId}`);
  }
}
